### PR TITLE
perf(cute): optimize SM100/SM103 hd256 long-sequence causal varlen prefill

### PR DESCRIPTION
**Pull request by muse-coder**
## Summary

Modern agent workloads increasingly operate on long Q/K/V sequences while also requiring variable-length batching and prefix-cache attention. The existing FA4 path is not yet optimized for this combination, leaving a significant efficiency gap and making attention one of the major bottlenecks in agent inference optimization. This PR adds a dedicated FlashAttention-4 forward kernel for causal `head_dim=256` attention on SM103, optimized for these long and irregular workloads. The goal is to make FA4 efficient and practical for the real-world agent workloads used by the community.

In production agent serving, many full-causal prefill attention workloads use FP8 Q/K/V tensors produced by directly casting BF16 activations to FP8. To support this common execution path efficiently, this PR also includes dedicated FP8 tuning for paged prefix-cache attention, rather than treating FP8 as an incidental extension of the BF16 kernel.

### Supported configurations

- BF16 and FP8 E4M3 Q/K/V inputs, with BF16 output
- MHA and GQA, including `pack_gqa`
- Regular and variable-length causal attention
- Contiguous and paged KV cache
- Exact per-request KV lengths through `seqused_k`
- KV-cache page sizes 16, 32, 64, 128, and 256
- `head_dim == head_dim_v == 256`
- Forward, `num_splits=1`

The specialized dispatch is currently enabled on SM103.

## Kernel design and optimizations

The kernel is specialized around the long-context HD256 workload instead of extending the generic forward path with another configuration.

- **2CTA pair-UMMA.** A `(2, 1)` CTA cluster processes an `M=256` tile as two 128-row Q tiles. Pair-UMMA (`cta_group::2`) is used for both QK and PV, giving the HD256 kernel enough work to use the Blackwell tensor cores efficiently.

- **CTA-local online softmax.** The accumulators are split along M, so each CTA owns its S, P, softmax state, and output rows. The two CTAs do not exchange row-max or row-sum values during online softmax.

- **TMA multicast for K/V.** K and V tiles are loaded once and multicast to the CTA pair, reducing duplicated global-memory traffic. For small physical KV pages, page-sized TMA transactions are aggregated into one logical KV tile.

- **K-direction ping-pong pipeline.** Separate S/P pipeline slots overlap `QK(i+1)` with softmax and PV work for block `i`, while keeping the K/V pipeline full across long sequences.

- **CLC persistent scheduling.** The hardware CLC scheduler dynamically assigns cluster tiles to resident CTA pairs. This improves load balancing for varlen batches and long-tail request distributions.

- **Warp-specialized execution.** Loading, pair-UMMA, online softmax/correction, and output epilogue work are assigned to separate warps and connected by asynchronous pipelines and mbarriers.

- **Datatype-specific tuning.** BF16 uses a `128x64` Q/K tile with up to five K/V stages to fit the SM103 shared-memory budget. FP8 uses a tuned `128x128` tile, compact register allocations, and SM103 hardware `ex2` behavior.

### Benchmark setup

- GPU: NVIDIA B300 (SM103)
- Attention: causal, `head_dim=256`, `num_splits=1`
- Head configurations: HQ16/HKV1 and HQ32/HKV2
- Dtypes: BF16 and FP8 E4M3 inputs; BF16 output
- Timing: 20 ms warmup, 100 ms measurement, 3 repeats; median reported
- Compilation, planning, input generation, and correctness checks are outside the timed region

The plots use the current FA4 implementation as the `1.0x` baseline:

```text
speedup over current FA4 = current FA4 latency / operator latency
```

A value above `1.0x` is faster than the current FA4 implementation. For the Page64 plots, the PR and FlashInfer `trtllm-gen` use paged KV while the current FA4 baseline is a contiguous-layout reference. The prefix-cache plot is a strict matched Page128 A/B with identical shapes, inputs, page tables, block size, and `pack_gqa=False` on both the PR and current FA4.

### Regular causal attention

Across representative Q=KV lengths from 1K to 256K, the PR reaches:

- BF16: `1.53x–1.64x` geometric-mean speedup over the current FA4 contiguous reference
- FP8 E4M3: `1.91x–2.13x` geometric-mean speedup
- Against FlashInfer `trtllm-gen` paged64, the full regular suite shows approximately `1.18x` for BF16 and `1.29x–1.30x` for FP8

Regular causal attention

### Variable-length causal attention

Across seven irregular B=4 batches per dtype/head configuration, the PR reaches:

- BF16: `1.31x–1.33x` geometric-mean speedup over the current FA4 contiguous reference
- FP8 E4M3: `1.83x–1.85x` geometric-mean speedup
- Against FlashInfer `trtllm-gen` paged64: `1.14x` for BF16 and `1.24x–1.25x` for FP8, with 28/28 wins

Variable-length causal attention

### Prefix-cache attention with matched block size

For FP8 prefix-cache cases with a large Q/KV gap, the strict Page128 matched A/B shows:

- HQ16/HKV1: `1.39x` geometric-mean speedup with `pack_gqa=False`
- HQ32/HKV2: `1.37x` geometric-mean speedup with `pack_gqa=False`
- 16/16 wins over the current FA4 implementation

Prefix-cache attention

## Testing

The added tests cover:

- BF16 and FP8 E4M3
- MHA and GQA
- Regular and asymmetric Q/KV lengths
- Contiguous and paged KV cache
- Exact logical KV lengths through `seqused_k`
- Random physical page mappings
- Page sizes 16, 32, 64, 128, and 256
- Numerical comparison against the PyTorch reference implementation

### CODE REVIEW

**Comment by muse-coder** [comment]
## Performance benchmark script

The following script was used to benchmark contiguous HD256 causal attention for the regular and variable-length results shown above. It intentionally excludes paged KV and prefix-cache cases. Run it once from the PR checkout and once from the current FA4 checkout, then compare rows with matching mode, dtype, head configuration, and sequence lengths.

```python
#!/usr/bin/env python3
"""Benchmark contiguous HD256 causal attention for regular and varlen shapes.

This script intentionally does not benchmark paged KV or prefix cache.
Run it once from the PR checkout and once from the upstream checkout, then
compare rows with the same mode, dtype, heads, and sequence lengths.
"""

from __future__ import annotations

import argparse
import inspect
import json
import platform
import statistics
import subprocess
import sys
import types
from pathlib import Path
from typing import Callable

import torch
import triton.testing

# Import FA4 directly from this checkout. The top-level flash_attn package also
# imports the separately built FA2 CUDA extension, which this benchmark does
# not use.
REPO_ROOT = Path(__file__).resolve().parents[1]
if "flash_attn" not in sys.modules:
    flash_attn_package = types.ModuleType("flash_attn")
    flash_attn_package.__path__ = [str(REPO_ROOT / "flash_attn")]
    sys.modules["flash_attn"] = flash_attn_package

from flash_attn.cute.interface import flash_attn_func, flash_attn_varlen_func


HEAD_DIM = 256
K = 1024
FP8 = torch.float8_e4m3fn

REGULAR_LENGTHS = (
    1 * K,
    2 * K,
    3 * K,
    4 * K,
    5 * K,
    6 * K,
    7 * K,
    8 * K,
    9 * K,
    10 * K,
    12 * K,
    16 * K,
    32 * K,
    64 * K,
    128 * K,
    256 * K,
)

VARLEN_GROUPS = (
    (1 * K, 2 * K, 3 * K, 4 * K),
    (4 * K, 6 * K, 8 * K, 10 * K),
    (8 * K, 10 * K, 12 * K, 16 * K),
    (16 * K, 24 * K, 32 * K, 48 * K),
    (32 * K, 48 * K, 64 * K, 96 * K),
    (64 * K, 96 * K, 128 * K, 192 * K),
    (128 * K, 160 * K, 192 * K, 256 * K),
)


def source_commit(source_file: Path) -> str | None:
    try:
        return subprocess.check_output(
            ["git", "-C", str(source_file.parents[2]), "rev-parse", "HEAD"],
            text=True,
        ).strip()
    except (subprocess.CalledProcessError, FileNotFoundError):
        return None


def build_cu_seqlens(lengths: tuple[int, ...]) -> torch.Tensor:
    values = [0]
    for length in lengths:
        values.append(values[-1] + length)
    return torch.tensor(values, device="cuda", dtype=torch.int32)


def make_input(
    shape: tuple[int, ...], dtype: torch.dtype, generator: torch.Generator
) -> torch.Tensor:
    tensor = torch.randn(
        shape, device="cuda", dtype=torch.bfloat16, generator=generator
    ).mul_(0.02)
    return tensor.to(dtype).contiguous()


def timed(
    fn: Callable[[], torch.Tensor], warmup_ms: int, rep_ms: int, repeats: int
) -> dict[str, float | list[float]]:
    fn()
    torch.cuda.synchronize()
    samples = [
        float(triton.testing.do_bench(fn, warmup=warmup_ms, rep=rep_ms))
        for _ in range(repeats)
    ]
    return {
        "samples_ms": samples,
        "median_ms": statistics.median(samples),
        "min_ms": min(samples),
        "max_ms": max(samples),
    }


def causal_flops(lengths: tuple[int, ...], hq: int) -> int:
    # QK and PV each require 2 FLOPs per multiply-add. Causal masking keeps
    # approximately half of the square attention matrix.
    return sum(2 * hq * length * length * HEAD_DIM for length in lengths)


def output_is_finite(out: torch.Tensor) -> bool:
    flat = out.flatten()
    stride = max(1, flat.numel() // (1024 * 1024))
    return bool(torch.isfinite(flat[::stride].float()).all())


def run_regular(
    length: int,
    *,
    hq: int,
    hkv: int,
    dtype: torch.dtype,
    pack_gqa: bool,
    warmup_ms: int,
    rep_ms: int,
    repeats: int,
) -> dict:
    generator = torch.Generator(device="cuda").manual_seed(
        20260724 + length + hq
    )
    q = make_input((1, length, hq, HEAD_DIM), dtype, generator)
    k = make_input((1, length, hkv, HEAD_DIM), dtype, generator)
    v = make_input((1, length, hkv, HEAD_DIM), dtype, generator)

    def run() -> torch.Tensor:
        out = flash_attn_func(
            q,
            k,
            v,
            causal=True,
            softmax_scale=HEAD_DIM**-0.5,
            num_splits=1,
            pack_gqa=pack_gqa,
        )
        return out[0] if isinstance(out, (tuple, list)) else out

    out = run()
    torch.cuda.synchronize()
    timing = timed(run, warmup_ms, rep_ms, repeats)
    flops = causal_flops((length,), hq)
    return {
        "record": "case",
        "mode": "regular",
        "batch": 1,
        "q_lengths": [length],
        "kv_lengths": [length],
        "hq": hq,
        "hkv": hkv,
        "head_dim": HEAD_DIM,
        "dtype": "fp8_e4m3" if dtype == FP8 else "bf16",
        "pack_gqa": pack_gqa,
        "finite": output_is_finite(out),
        "timing": timing,
        "tflops": flops / (timing["median_ms"] * 1e-3) / 1e12,
    }


def run_varlen(
    lengths: tuple[int, ...],
    *,
    hq: int,
    hkv: int,
    dtype: torch.dtype,
    pack_gqa: bool,
    warmup_ms: int,
    rep_ms: int,
    repeats: int,
) -> dict:
    generator = torch.Generator(device="cuda").manual_seed(
        20260724 + sum(lengths) + hq
    )
    total = sum(lengths)
    q = make_input((total, hq, HEAD_DIM), dtype, generator)
    k = make_input((total, hkv, HEAD_DIM), dtype, generator)
    v = make_input((total, hkv, HEAD_DIM), dtype, generator)
    cu_seqlens = build_cu_seqlens(lengths)

    def run() -> torch.Tensor:
        out = flash_attn_varlen_func(
            q,
            k,
            v,
            cu_seqlens_q=cu_seqlens,
            cu_seqlens_k=cu_seqlens,
            max_seqlen_q=max(lengths),
            max_seqlen_k=max(lengths),
            causal=True,
            softmax_scale=HEAD_DIM**-0.5,
            num_splits=1,
            pack_gqa=pack_gqa,
        )
        return out[0] if isinstance(out, (tuple, list)) else out

    out = run()
    torch.cuda.synchronize()
    timing = timed(run, warmup_ms, rep_ms, repeats)
    flops = causal_flops(lengths, hq)
    return {
        "record": "case",
        "mode": "varlen",
        "batch": len(lengths),
        "q_lengths": list(lengths),
        "kv_lengths": list(lengths),
        "hq": hq,
        "hkv": hkv,
        "head_dim": HEAD_DIM,
        "dtype": "fp8_e4m3" if dtype == FP8 else "bf16",
        "pack_gqa": pack_gqa,
        "finite": output_is_finite(out),
        "timing": timing,
        "tflops": flops / (timing["median_ms"] * 1e-3) / 1e12,
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--label", default="fa4")
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument(
        "--mode", choices=("regular", "varlen", "both"), default="both"
    )
    parser.add_argument(
        "--dtype", choices=("bf16", "fp8", "both"), default="both"
    )
    parser.add_argument(
        "--heads",
        action="append",
        choices=("16:1", "32:2"),
        help="May be repeated; defaults to both head configurations",
    )
    parser.add_argument(
        "--pack-gqa", choices=("true", "false"), default="false"
    )
    parser.add_argument("--warmup-ms", type=int, default=20)
    parser.add_argument("--rep-ms", type=int, default=100)
    parser.add_argument("--repeats", type=int, default=3)
    args = parser.parse_args()

    torch.cuda.set_device(0)
    modes = ("regular", "varlen") if args.mode == "both" else (args.mode,)
    dtypes = ("bf16", "fp8") if args.dtype == "both" else (args.dtype,)
    heads = args.heads or ("16:1", "32:2")
    pack_gqa = args.pack_gqa == "true"
    interface_path = Path(inspect.getsourcefile(flash_attn_varlen_func)).resolve()

    metadata = {
        "record": "metadata",
        "label": args.label,
        "gpu": torch.cuda.get_device_name(0),
        "compute_capability": list(torch.cuda.get_device_capability(0)),
        "gpu_uuid": str(torch.cuda.get_device_properties(0).uuid),
        "python": platform.python_version(),
        "torch": torch.__version__,
        "interface_path": str(interface_path),
        "source_commit": source_commit(interface_path),
        "head_dim": HEAD_DIM,
        "causal": True,
        "kv_layout": "contiguous",
        "pack_gqa": pack_gqa,
        "warmup_ms": args.warmup_ms,
        "rep_ms": args.rep_ms,
        "repeats": args.repeats,
        "argv": sys.argv,
    }

    args.output.parent.mkdir(parents=True, exist_ok=True)
    with args.output.open("w", encoding="utf-8") as output:
        output.write(json.dumps(metadata, sort_keys=True) + "\n")
        output.flush()
        print(json.dumps(metadata, sort_keys=True), flush=True)

        for dtype_name in dtypes:
            dtype = torch.bfloat16 if dtype_name == "bf16" else FP8
            for head_spec in heads:
                hq, hkv = (int(value) for value in head_spec.split(":"))
                if "regular" in modes:
                    for length in REGULAR_LENGTHS:
                        row = run_regular(
                            length,
                            hq=hq,
                            hkv=hkv,
                            dtype=dtype,
                            pack_gqa=pack_gqa,
                            warmup_ms=args.warmup_ms,
                            rep_ms=args.rep_ms,
                            repeats=args.repeats,
                        )
                        output.write(json.dumps(row, sort_keys=True) + "\n")
                        output.flush()
                        print(json.dumps(row, sort_keys=True), flush=True)
                if "varlen" in modes:
                    for lengths in VARLEN_GROUPS:
                        row = run_varlen(
                            lengths,
                            hq=hq,
                            hkv=hkv,
                            dtype=dtype,
                            pack_gqa=pack_gqa,
                            warmup_ms=args.warmup_ms,
                            rep_ms=args.rep_ms,
                            repeats=args.repeats,
                        )
                        output.write(json.dumps(row, sort_keys=True) + "\n")
                        output.flush()
                        print(json.dumps(row, sort_keys=True), flush=True)


if __name__ == "__main__":
    main()
```
